pageserver/src/consumption_metrics.rs

```diff
@@ -273,31 +278,48 @@ pub async fn collect_metrics_iteration(
         })
         .expect("PageserverConsumptionMetric should not fail serialization");
 
-        let res = client
-            .post(metric_collection_endpoint.clone())
-            .json(&chunk_json)
-            .send()
-            .await;
-
-        match res {
-            Ok(res) => {
-                if res.status().is_success() {
-                    // update cached metrics after they were sent successfully
-                    for (curr_key, curr_val) in chunk.iter() {
-                        cached_metrics.insert(curr_key.clone(), *curr_val);
-                    }
-                } else {
-                    error!("metrics endpoint refused the sent metrics: {:?}", res);
-                    for metric in chunk_to_send.iter() {
-                        // Report if the metric value is suspiciously large
-                        if metric.value > (1u64 << 40) {
+        const MAX_RETRIES: u32 = 3;
+
+        for attempt in 0..MAX_RETRIES {
+            let started_at = Instant::now();
+            let res = client
+                .post(metric_collection_endpoint.clone())
+                .json(&chunk_json)
+                .send()
+                .await;
+
+            match res {
+                Ok(res) => {
+                    if res.status().is_success() {
+                        // update cached metrics after they were sent successfully
+                        for (curr_key, curr_val) in chunk.iter() {
+                            cached_metrics.insert(curr_key.clone(), *curr_val);
+                        }
+                    } else {
+                        error!("metrics endpoint refused the sent metrics: {:?}", res);
+                        for metric in chunk_to_send
+                            .iter()
+                            .filter(|metric| metric.value > (1u64 << 40))
+                        {
+                            // Report if the metric value is suspiciously large
                             error!("potentially abnormal metric value: {:?}", metric);
                         }
                     }
+                    break;
+                }
+                Err(err) if err.is_timeout() => {
+                    error!(attempt, "timeout sending metrics, retrying immediately");
+                    crate::tenant::tasks::warn_when_period_overrun(
+                        started_at.elapsed(),
+                        DEFAULT_HTTP_REPORTING_TIMEOUT,
+                        "consumption_metrics",
+                    );
```

Comment:
Ah, sorry, my last review comment was ambiguous.

I don't think it makes sense to warn for period overruns in this retry loop.

What I meant is to detect `collect_metrics_iteration` overruns.

Move this block to the `collect_metrics` function, where we return from `collect_metrics_iteration`.
And use `metric_collection_interval` instead of the `DEFAULT_HTTP_REPORTING_TIMEOUT` as the period argument.

The resulting behavior will be that, if a single `collect_metrics_iteration` takes longer than `metric_collection_interval`, we'll print a warning.

Also, we need the same thing for `calculate_synthetic_size_worker`.

If you think it's too much work in this PR, feel free to back out all `warn_when_period_overrun` code, I can take it on in a separate PR.

I think they fit into the overall theme of this PR though.